File: apps/holefeeder-mobile/src/flows/core/flows/transaction.ts
import { CategoryType } from '@/flows/core/categories/category-type';
import { DateOnly } from '@/shared/core/date-only';
import { Id } from '@/shared/core/id';
import { Money } from '@/shared/core/money';
import { TagList } from './tag-list';

export type Transaction = {
  id: Id;
  date: DateOnly;
  amount: Money;
  description: string;
  accountId: Id;
  categoryId: Id;
  categoryType: CategoryType;
  tags: TagList;
  cashflowId?: Id;
  cashflowDate?: DateOnly;
};

const valid = (value: Record<string, unknown>): Transaction => ({
  id: Id.valid(value.id),
  date: DateOnly.valid(value.date),
  amount: Money.valid(value.amount),
  description: value.description as string,
  accountId: Id.valid(value.accountId),
  categoryId: Id.valid(value.categoryId),
  categoryType: CategoryType.valid(value.categoryType),
  tags: TagList.valid(value.tags),
  cashflowId: value.cashflowId ? Id.valid(value.cashflowId) : undefined,
  cashflowDate: value.cashflowDate ? DateOnly.valid(value.cashflowDate) : undefined,
});

export const Transaction = { valid: valid };

File: backend/src/Holefeeder.Domain/Features/Transactions/Transaction.cs
using Holefeeder.Domain.Features.Accounts;
using Holefeeder.Domain.Features.Categories;
using Holefeeder.Domain.Features.Users;
using Holefeeder.Domain.ValueObjects;

namespace Holefeeder.Domain.Features.Transactions;

public partial record Transaction : IAggregateRoot
{
    private IReadOnlyCollection<string> _tags = [];

    private Transaction(TransactionId id, DateOnly date, Money amount, AccountId accountId, CategoryId categoryId, UserId userId)
    {
        Id = id;
        Date = date;
        Amount = amount;
        AccountId = accountId;
        CategoryId = categoryId;
        UserId = userId;
    }

    public TransactionId Id { get; private init; }

    public DateOnly Date { get; private init; }

    public Money Amount { get; private init; }

    public string Description { get; init; } = string.Empty;

    public AccountId AccountId { get; init; }

    public Account? Account { get; init; }

    public CategoryId CategoryId { get; init; }

    public Category? Category { get; init; }

    public CashflowId? CashflowId { get; private init; }

    public Cashflow? Cashflow { get; init; }

    public DateOnly? CashflowDate { get; private init; }

    public IReadOnlyCollection<string> Tags => _tags.ToImmutableList();

    public UserId UserId { get; init; }
}

public sealed record TransactionId : StronglyTypedId<TransactionId>;
